### ISSUE TITLE
GoTest does not handle errors in sub-packages

### ISSUE THREAD

**Issue by raphael** [fatih/vim-go#916]
### Actual behavior

GoTest does not handle errors happening in files located in a sub-directory, instead vim-go prints:

```
vim-go: [run] Dropped "/home/raphael/go/src/github.com/18-goa/registration_testing.go" from location list (nonvalid filename) 
```

The file is actually in:

```
"/home/raphael/go/src/github.com/18-goa/app/test/registration_testing.go
```

But the test function lives in:

```
"/home/raphael/go/src/github.com/18-goa/registration_test.go"
```

### Expected behavior

GoTest should list all the errors in the `QuickFix` window. If possible it should also link to the right file.

Here is the output from `go test`:

```
--- FAIL: TestCreateRegistrationBadRequest (0.00s)
        registration_testing.go:79: invalid response status code: got 201, expected 400
        registration_testing.go:86: invalid response media: got &{Address:0xc8201dcf00 FirstName:a Href:/api/registrations/21 ID:21 LastName:last Links:0xc82002c318}, expected instance of goa.Error
        registration_test.go:139: short-firstname: reponse media type is nil
        registration_testing.go:79: invalid response status code: got 201, expected 400
        registration_testing.go:86: invalid response media: got &{Address:0xc8201dd340 FirstName:first Href:/api/registrations/22 ID:22 LastName:a Links:0xc82002c320}, expected instance of goa.Error
        registration_test.go:139: short-lastname: reponse media type is nil
--- FAIL: TestListRegistrationOK (0.00s)
        registration_testing.go:374: controller returned index error: argument must be a string: (*int)(nil), logs:
FAIL
exit status 1
FAIL    github.com/18-goa       0.275s
```
As you can see it doesn't provide the path to the files, just the file names. It would be awesome if vim-go could have a little heuristic where it searches in the sub-directories if it can't find the file in the current directory instead of failing like the above.

### Steps to reproduce:

Create tests that use a sub-package which fails the tests.

### Configuration

* vim-go version: latest
* go version: 1.6.2 (latest)

**Comment by bruschill** [comment]
I'm having this same issue. 

<img width="1440" alt="screen shot 2016-06-29 at 11 38 58 am" src="https://cloud.githubusercontent.com/assets/621389/16460676/251b466a-3dee-11e6-9585-6c5fffb57abb.png">

I'm not sure if it's where the root of issue is, but it doesn't look like any recent change to [`go#tool#ParseErrors`](https://github.com/fatih/vim-go/blob/1a9cc786751947441f05b92f380540dee388feff/autoload/go/tool.vim#L46) would have caused it.

@fatih - If you can give me a few ideas of where the problem could be I'll gladly dive in and see if I can come up with a fix.

**Comment by fatih** [comment]
Hi @raphael @bruschill 

Can you please provide a test that use a sub-package which fails the tests? I didn't undestand how you call the test inside the sub-package, from the upper level test? Do you call something like `:GoTest ./...` ?

**Comment by bruschill** [comment]
@fatih I can't speak for @raphael, but I call `:GoTest -bench .`.

**Comment by fatih** [comment]
@bruschill could you please provide the necessary files to reproduce it? I really have hard time following what the problem is from the comment. I just want to know how the files are laid out with the sample files, like `foo.go`, `foo_test.go`, etc.. so I can reproduce and fix the issue.

**Comment by fatih** [comment]
@bruschill btw, I have an idea, but I have to be able to reproduce it.

**Comment by bruschill** [comment]
@fatih Absolutely. The project files are [here](https://dl.dropboxusercontent.com/u/8137552/gigasecond.zip). The project is from the free programming language learning site Exercism.

**Comment by fatih** [comment]
@bruschill the problem you have is different than what @raphael has. You have a `t.Errorf`, `t.Fatal` that is multiline so it's hard to parse it. It's dropping the other lines as they are not valids but jumps to the correct file. Please open another issue if you think it's wrong or if it doesn't jump to the error.

**Comment by bruschill** [comment]
@fatih my mistake. I'll submit a separate issue in a bit. Thanks for your help!

**Comment by raphael** [comment]
@fatih thanks for looking into it, here's a repro: https://github.com/raphael/vim-go-test-issue

**Comment by fatih** [comment]
@raphael  alright ,this is the first time I see someone importing a test and using it :) I didn't know that there was a case for this. Is this a case people use a lot? Why is the test inside another package?

**Comment by bradleyfalzon** [comment]
@fatih hashicorp talk about providing helpers in their recent talk: https://speakerdeck.com/mitchellh/advanced-testing-with-go?slide=53 / https://youtu.be/yszygk1cpEc (53:56)

Although mostly to provide helpers, some of the functions also do execute t.Fatal/t.Error themselves, eg: https://github.com/hashicorp/consul/blob/master/testutil/wait.go

**Comment by raphael** [comment]
Yeah that's the exact use case: test helpers. These are not part of the test per se but they do accept `t` as argument and may make the tests fail it they can't setup / mock / do whatever it is they are supposed to be doing that's always supposed to work.

**Comment by fatih** [comment]
Thanks all for the info. This is enough for me to fix the problem. The main problem is that `go test` prints file names relative to the `package`. So here is the problem:

```
➜  vim-go-test-issue git:(master) pwd
/Users/fatih/Code/go/src/github.com/raphael/vim-go-test-issue
➜  vim-go-test-issue git:(master) go test
--- FAIL: TestThis (0.00s)
	fail.go:6: Whoops - I am not listed in vim's QuickFix window
FAIL
exit status 1
FAIL	github.com/raphael/vim-go-test-issue	0.006s
```

So where is this `fail.go` file now? Exactly you don't know it! Because it just prints the filename relative to the `package`. So it's extremely difficult to solve it unfortunately as there is no easy way to know where the file residues. Even `-v` doesn't print it. I'm going to look if there is any kind of info we can gather, if not I'm not sure if there is an easy way to solve it (without adding complex logic to `cmd.vim`)

**Comment by raphael** [comment]
The initial description was saying as much :)

> As you can see it doesn't provide the path to the files, just the file names. It would be awesome if vim-go could have a little heuristic where it searches in the sub-directories if it can't find the file in the current directory instead of failing like the above.

I'm suggesting `vim-go` could do a best effort and look in all sub-directories - recursively - for a file with the same name. Yes there may be false positive and yes the file might not be there but in reality it will work 90% of the case and when it doesn't it won't be worse than the current behavior.

**Comment by raphael** [comment]
Also a quick note that the "false positive" problem exists today: if a file in a sub-package has the same filename as a file in the current directory then vim-go ends up using the file in the current directory instead of the one in the sub-package. That got me confused a few times (understand there isn't much you can do about it).

**Comment by raphael** [comment]
I understand - that's why I noted "the file might not be there". Still in all the cases that I've seen the helpers are in a sub-directory of the package being tested. Agreed that it would be better for the `go` tool itself to list the relative path instead of just filenames.

**Comment by fatih** [comment]
I'm closing this as there is not much we can do on vim-go side. I don't want to implement custom logic for this case. Unfortunately until `go test` itself provides as more information there is not much we can do. I hope you all understand this. Happy to merge any fixes though that does this in an elegant way. Thanks.
